### PR TITLE
Add max_age / improve prompt=login - now without grace periods! 

### LINKED ISSUE #4321: Issues with prompt / login_hint
### I did this

I am trying to connect stalwart mail server webui to kanidm. 
Stalwart unfortunately always adds prompt=login&login_hint=user@example.com to the url, where user@example.com is the mail address and not my kanidm user.

### I expected the following

- change username to user
- login with passkey
- redirect to stalwart

### This happened instead

- change username to user
- login with passkey
- redirect to login page (Username prompt) on URL /ui/oauth2/resume
- change username to user
- login with passkey
- redirect to login page (Username prompt) on URL /ui/oauth2/resume
- ...

### Kanidm version details

- Output of `kanidm(d) version`: kanidm 1.10.1
- Are you running it in a container? If so, which image/tag?: yes 1.10.1

### Any other comments

As I understand it, login_hint should be a suggestion, not a restriction. But even if it was, there should be an error message.

After removing login_hint/prompt from the url by patching the stalwart webui, login works.

### PR DESCRIPTION
**Pull request by Firstyear**
This improves prompt=login, importantly cycle breaking when the login is performed from the session resumption.

Fixes #4321

Checklist

- [x] This PR contains no AI generated code
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

**Comment by grisia** [comment]
Please excuse me, but I think I would like to briefly raise a concern here.

I am not sure it should be Kanidm's responsibility to decide whether an authentication is "fresh enough". It seems to me that only the client should decide whether it wants to require a fresh authentication.

The current implementation might introduce a security concern, because within that window no re-authentication is required. It also seems possible that a user may intentionally want to sign in as a different account. In that case, would they have to wait 300 seconds before being able to proceed?

The closed PR:https://github.com/kanidm/kanidm/pull/4332 I proposed, although it may not be the most elegant solution, takes a different approach: it keeps track of whether authentication has just been completed successfully in the current flow. Only in that case is the flow allowed to continue.

**Comment by Firstyear** [comment]
> Please excuse me, but I think I would like to briefly raise a concern here.

No need to feel sorry, when we make decisions we need to be able to explain and justify them. That's part of a good project.

> I am not sure it should be Kanidm's responsibility to decide whether an authentication is "fresh enough". It seems to me that only the client should decide whether it wants to require a fresh authentication.

I think that it is very much within Kanidm's responsibility as we are the identity provider, charged with making security decisions that clients (oidc relying parties) rely upon. 

Now more importantly in this case prompt=login is not a security control, it is a UI guidance control. This can be easily demonstrated - if prompt=login is removed from the GET query, then the RP has no signal or knowledge that it was removed. 

The strict version of this is the "max_age" parameter which *does* enforce a time window for freshness of the session - exactly the same as what we are doing in our code here. There is a max_age=0 but that has it's own issues. Additionally, if max_age is set, then the IDP must add the auth_time claim - today we don't actually support this flow however, but we should. 

>
> The current implementation might introduce a security concern, because within that window no re-authentication is required. It also seems possible that a user may intentionally want to sign in as a different account. In that case, would they have to wait 300 seconds before being able to proceed?

The current implementation of prompt=login is using our existing auth code, not the re-auth code. That's why it falsely presents you with the ability to change accounts. It shouldn't do that. 

I think looking at it, we may have released prompt=login support too soon as it seems to be causing more issues for users, as well as confusion, and logically it has some mistakes that interact with things like max_age. 

I'm going to propose it is disabled in 1.10 so that we can improve it in 1.11.

**Comment by Firstyear** [comment]
@grisia We have accepted the removal of prompt=login for 1.10 in the meantime. The update to this PR takes some inspiration from your PR, but goes further to improve the reauth session handling and validation.

**Comment by Firstyear** [comment]
@grisia You are also welcome to review, I included ideas from your PR into this.

**Comment by grisia** [comment]
This looks good to me overall and matches what I would expect from `prompt=login`.

The one edge case I hit in a live test was being already signed into Kanidm as one user, then opening the Stalwart admin UI as another. With a mismatched `login_hint`, I think a fresh login flow is more correct than reauthenticating the existing session.

This also feels adjacent to the multi-session/account-switching discussion in #2824.

I opened a small follow-up for that case here:
https://github.com/Firstyear/kanidm/pull/1

**Comment by Firstyear** [comment]
> The one edge case I hit in a live test was being already signed into Kanidm as one user, then opening the Stalwart admin UI as another. With a mismatched `login_hint`, I think a fresh login flow is more correct than reauthenticating the existing session.

We don't support login's with email addresses today, so we have no way to know if the login_hint mismatches to the account. I think that will have to be a follow up issue as there are other changes to the login flow that would need to occur to make that work correctly.

**Comment by grisia** [comment]
Just to clarify: this is not about email-based login. I am aware Kanidm does not support that today, so in Stalwart I already only use the plain login name as `login_hint`.

The behavior that still feels odd to me is when I already have a Kanidm session as Alice, but a client sends `prompt=login` with `login_hint=bob`, and Kanidm then reauthenticates Alice.

For matching hints, reauth seems correct.
For a different hinted login name, I would expect Kanidm to fall back to a fresh authentication flow instead.

In the follow-up I tried to correct that behavior, and in my local setup it behaves as expected with that change.

File: server/core/src/https/views/oauth2.rs [hunk 1]
```diff
@@ -90,36 +103,46 @@ async fn oauth2_auth_req(
     client_auth_info: ClientAuthInfo,
     domain_info: DomainInfoRead,
     jar: CookieJar,
-    maybe_auth_req: Option<AuthorisationRequest>,
+    // maybe_auth_req: Option<AuthorisationRequest>,
```
Comment (yaleman) [inline server/core/src/https/views/oauth2.rs]:
```suggestion
```

File: server/lib/src/constants/mod.rs [hunk 2]
```diff
@@ -191,6 +191,9 @@ pub const OAUTH_REFRESH_TOKEN_EXPIRY: u32 = 3600 * 16;
 /// of the refresh token, which is bound to the issuing session.
 pub const OAUTH2_ACCESS_TOKEN_EXPIRY: u32 = 15 * 60;
 
+/// The absolute maximum that can be requested for max-age requests in OIDC.
+pub const OAUTH2_OIDC_MAX_AGE_CLAMP: i64 = 86400;
```
Comment (yaleman) [inline server/lib/src/constants/mod.rs:195]: why don't we just alias this to `DEFAULT_AUTH_SESSION_EXPIRY` so that we don't end up with a mismatch later?

**Review by yaleman — CHANGES_REQUESTED** [review]
Minor questions, yay for dumping the grace period which seemed like a bad pattern.

File: server/lib/src/constants/mod.rs (hunk 2 above)
Comment (Firstyear) [reply server/lib/src/constants/mod.rs:195]: Because the auth session length could be extended by account policy. I think the max_age here should be detached from that though, really this clamp is just here to ensure that max_age isn't set to a ridiculous value.

**Review by yaleman — APPROVED** [review]
